src/geometry.rs

```diff
@@ -0,0 +1,137 @@
+//! Structures for creating and storing geometric primitives.
+use mint;
+use std::collections::HashMap;
+use genmesh::{Vertex as GenVertex, EmitTriangles, Triangulate};
+use genmesh::generators::{self, IndexedPolygon, SharedVertex};
+
+/// A shape of geometry that is used for mesh blending.
+#[derive(Clone, Debug)]
+pub struct Shape {
+    /// Vertices.
+    pub vertices: Vec<mint::Point3<f32>>,
+    /// Normals.
+    pub normals: Vec<mint::Vector3<f32>>,
+    /// Tangents.
+    pub tangents: Vec<mint::Vector4<f32>>,
+    /// Texture co-ordinates.
+    pub tex_coords: Vec<mint::Point2<f32>>,
+}
+
+impl Shape {
+    /// Create an empty shape.
+    pub fn empty() -> Self {
+        Shape {
+            vertices: Vec::new(),
+            normals: Vec::new(),
+            tangents: Vec::new(),
+            tex_coords: Vec::new(),
+        }
+    }
+}
+
+/// A collection of vertices, their normals, and faces that defines the
+/// shape of a polyhedral object.
+#[derive(Clone, Debug)]
+pub struct Geometry {
+    /// The original shape of geometry.
+    pub base_shape: Shape,
+    /// A map containing blend shapes and their names.
+    pub shapes: HashMap<String, Shape>,
+    /// Faces.
+    pub faces: Vec<[u32; 3]>,
+}
+
+impl Geometry {
+    /// Create new `Geometry` without any data in it.
+    pub fn empty() -> Self {
+        Geometry {
+            base_shape: Shape::empty(),
+            shapes: HashMap::new(),
+            faces: Vec::new(),
+        }
+    }
+
+    /// Create `Geometry` from vector of vertices.
+    pub fn with_vertices(vertices: Vec<mint::Point3<f32>>) -> Self {
+        Geometry {
+            base_shape: Shape {
+                vertices,
+                normals: Vec::new(),
+                .. Shape::empty()
+            },
+            .. Geometry::empty()
+        }
+    }
+
+    fn generate<P, G, Fpos, Fnor>(gen: G, fpos: Fpos, fnor: Fnor) -> Self where
+        P: EmitTriangles<Vertex=usize>,
+        G: IndexedPolygon<P> + SharedVertex<GenVertex>,
+        Fpos: Fn(GenVertex) -> mint::Point3<f32>,
```

Comment:
this is not for your PR, but later on this would probably be cleaner and faster with a single function returning both position and normal